In [1]:
pip install ydata-profiling

Note: you may need to restart the kernel to use updated packages.


In [2]:
import numpy as np
import pandas as pd
from ydata_profiling import ProfileReport
from urllib.parse import urlparse
from functions_url import *

/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [3]:
import matplotlib.pyplot as plt
import seaborn as sns

In [4]:
df = pd.read_csv("dataset_pishing.csv", sep = ",")

In [5]:
df.head(10)

url      status
0              http://www.crestonwood.com/router.php  legitimate
1  http://shadetreetechnology.com/V4/validation/a...    phishing
2  https://support-appleld.com.secureupdate.duila...    phishing
3                                 http://rgipt.ac.in  legitimate
4  http://www.iracing.com/tracks/gateway-motorspo...  legitimate
5                   http://appleid.apple.com-app.es/    phishing
6                                http://www.mutuo.it  legitimate
7  http://www.shadetreetechnology.com/V4/validati...    phishing
8         http://vamoaestudiarmedicina.blogspot.com/  legitimate
9  https://parade.com/425836/joshwigler/the-amazi...  legitimate

In [6]:
df.shape

(11430, 2)

In [7]:
df.describe()

url      status
count                                               11430       11430
unique                                              11429           2
top     http://e710z0ear.du.r.appspot.com/c:/users/use...  legitimate
freq                                                    2        5715

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11430 entries, 0 to 11429
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   url     11430 non-null  object
 1   status  11430 non-null  object
dtypes: object(2)
memory usage: 178.7+ KB


In [9]:
df.status.value_counts()

status
legitimate    5715
phishing      5715
Name: count, dtype: int64

In [10]:
df['url'].value_counts()

url
http://e710z0ear.du.r.appspot.com/c:/users/user/downlo                                                                                                                                                                                                                                                                                                                                                                                                                                           2
http://www.crestonwood.com/router.php                                                                                                                                                                                                                                                                                                                                                                                                                                                            1
https://www.riverbed.com/produ

In [11]:
df['url'].mode().iloc[0]

'http://e710z0ear.du.r.appspot.com/c:/users/user/downlo'

In [12]:
df[df['url'] == 'http://e710z0ear.du.r.appspot.com/c:/users/user/downlo']

url    status
6035   http://e710z0ear.du.r.appspot.com/c:/users/use...  phishing
11305  http://e710z0ear.du.r.appspot.com/c:/users/use...  phishing

1. Cargue el dataset en un dataframe de pandas, muestre un ejemplo de cinco observaciones.
   -No tiene valores nulos, asi que la data esta limpia
   -Solo se clasifica el status en phishing o en legitimate
   -Tenemos la misma cantidad de legitimos y de phishing
   -El valor "http://e710z0ear.du.r.appspot.com/c:/users/user/downlo" esta repetido dos veces.
   -Tienen el mismo status el valor repetido.
2. Muestre la cantidad de observaciones etiquetadas en la columna status como “legit” y como
“pishing”. ¿Está balanceado el dataset?
    -Claro como mencionaba en el inciso anterior si esta balanceado porque hay la misma cantidad de ambos. aunque tecnicamnete hay un dato       repetido. por lo que si nos ponemos muy tecnicos esta "relativamente balanceado" pero en general para mi esta muy bien. de ser      necesario lo balancearemos quitando el valor extra y uno de los de legitimate.
    

Derivación de características

1. ¿Qué ventajas tiene el análisis de una URL contra el análisis de otros datos, cómo el tiempo
de vida del dominio, o las características de la página Web?

Que podes identificar solo con ese texto mucha información relevante sin necesidad de gastar recursos masivos, ademas se peude aprovechar el analisis de texto para ver caracteres inusuales que no se ven normalmente. Tambien se puede analizar un numero extenso de urls ya que usa pocos recursos. Tambien hay sitios webs que van aumentando los casos y se puede aprovechar esa data en tiempo real para saber que URLs son las mas conocidas para eso.
   
2. ¿Qué características de una URL son más prometedoras para la detección de phishing?
   
la presencia de subdominios, nombres de dominio, rutas de archivos y parámetros. Además, características como el número total de barras, la longitud, las posiciones de los puntos y los números en los nombres de subdominios y URL son importantes. Otras características valiosas son los componentes sintácticos de la URL, que son jerárquicos en secuencia, y el uso de algoritmos metaheurísticos y modelos de aprendizaje automático para fines de entrenamiento y detección.






En base a la respuesta anterior escriba al menos quince funciones basadas en los artículos, para
derivar características que un modelo pueda utilizar y añada dichas características al dataset original.

-Función para extraer el número de subdominios en una URL.
-Función para calcular la longitud del nombre de dominio.
-Función para determinar la presencia de rutas de archivos en la URL.
-Función para contar el número de parámetros en la URL.
-Función para calcular el número total de barras en la URL.
-Función para identificar las posiciones de los puntos en la URL.
-Función para detectar números en el subdominio y los nombres de URL.
-Función para analizar la estructura jerárquica de la URL.
-Función para extraer la extensión del dominio (por ejemplo, .com, .org).
-Función para verificar la presencia de caracteres poco comunes en la URL.
-Función para determinar la presencia de guiones en el nombre de dominio.
-Función para calcular la proporción de vocales a consonantes en la URL.
-Función para identificar el uso de caracteres especiales en la URL.
-Función para analizar la entropía de los componentes de la URL.
-Función para detectar patrones en secuencias de URL.

Preprocesamiento

Pasar a valores binarios

In [13]:
maping = {'legitimate': 0, 'phishing': 1}
# Convertir los valores en la columna 'status' a números
df['status_numerico'] = df['status'].replace(maping)


In [14]:
df.head()

url      status  \
0              http://www.crestonwood.com/router.php  legitimate   
1  http://shadetreetechnology.com/V4/validation/a...    phishing   
2  https://support-appleld.com.secureupdate.duila...    phishing   
3                                 http://rgipt.ac.in  legitimate   
4  http://www.iracing.com/tracks/gateway-motorspo...  legitimate   

   status_numerico  
0                0  
1                1  
2                1  
3                0  
4                0

In [15]:
df.drop(columns=['status'], inplace=True)

In [16]:
##Apply all the new columns from the functions
df['Subdomain_Count'] = df['url'].apply(extract_subdomain_count)
df['Domain_Length'] = df['url'].apply(calculate_domain_length)
df['Has_File_Path'] = df['url'].apply(has_file_path)
df['Parameter_Count'] = df['url'].apply(count_parameters)
df['Slash_Count'] = df['url'].apply(count_slashes)
df['Dot_Positions'] = df['url'].apply(identify_dots_positions)
df['Number_Count'] = df['url'].apply(detect_numbers)
df['Hierarchy_Level'] = df['url'].apply(analyze_hierarchy)
df['Domain_Extension'] = df['url'].apply(extract_domain_extension)
df['Has_Uncommon_Chars'] = df['url'].apply(check_uncommon_characters)
df['Has_Hyphens_Domain'] = df['url'].apply(check_hyphens_domain)
df['Vowel_Consonant_Ratio'] = df['url'].apply(calculate_vowel_consonant_ratio)
df['Special_Char_Count'] = df['url'].apply(identify_special_characters)
df['Entropy'] = df['url'].apply(analyze_entropy)
df['Has_Url_Pattern'] = df['url'].apply(detect_url_patterns)

In [17]:
df.head()

url  status_numerico  \
0              http://www.crestonwood.com/router.php                0   
1  http://shadetreetechnology.com/V4/validation/a...                1   
2  https://support-appleld.com.secureupdate.duila...                1   
3                                 http://rgipt.ac.in                0   
4  http://www.iracing.com/tracks/gateway-motorspo...                0   

   Subdomain_Count  Domain_Length  Has_File_Path  Parameter_Count  \
0                2             19              1                1   
1                1             23              1                1   
2                4             50              1                3   
3                2             11              0                1   
4                2             15              1                1   

   Slash_Count     Dot_Positions  Number_Count  Hierarchy_Level  \
0            3      [10, 22, 33]             0                2   
1            5              [26]            17                4   
2            5  [23, 27, 40, 54]            19                4   
3            2          [12, 15]             0                1   
4            5          [10, 18]             0                4   

  Domain_Extension  Has_Uncommon_Chars  Has_Hyphens_Domain  \
0              com               False               False   
1              com                True               False   
2              com                True                True   
3               in               False               False   
4              com               False               False   

   Vowel_Consonant_Ratio  Special_Char_Count   Entropy  Has_Url_Pattern  
0               0.363636                   0  0.666667                1  
1               0.827586                   1  0.666667                1  
2               0.517241                   1  0.833333                1  
3               0.300000                   0  0.500000                1  
4               0.363636                   0  0.666667                1

Visualización de los resultados

In [19]:
profile = ProfileReport(df, title="Profiling Report")

In [20]:
profile.to_file("your_report.html")

Export report to file: 100%|█████████████████████| 1/1 [00:00<00:00, 175.19it/s]


In [21]:
##Seleccion de caracteristicas

In [22]:
df.drop(columns=['Has_Url_Pattern', 'Dot_Positions','Domain_Length', 'Domain_Extension','Vowel_Consonant_Ratio', 'Has_Hyphens_Domain', 'url'], inplace=True)

In [23]:
##RANDOM FOREST

In [24]:
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score


In [25]:
df

status_numerico  Subdomain_Count  Has_File_Path  Parameter_Count  \
0                    0                2              1                1   
1                    1                1              1                1   
2                    1                4              1                3   
3                    0                2              0                1   
4                    0                2              1                1   
...                ...              ...            ...              ...   
11425                0                2              1                1   
11426                1                2              1                1   
11427                0                2              1                1   
11428                0                2              1                1   
11429                1                3              1               10   

       Slash_Count  Number_Count  Hierarchy_Level  Has_Uncommon_Chars  \
0                3             0                2               False   
1                5            17                4                True   
2                5            19                4                True   
3                2             0                1               False   
4                5             0                4               False   
...            ...           ...              ...                 ...   
11425            4             0                3               False   
11426            5             2                4                True   
11427            5            15                4                True   
11428            3             0                2               False   
11429            4            41                3                True   

       Special_Char_Count   Entropy  
0                       0  0.666667  
1                       1  0.666667  
2                       1  0.833333  
3                       0  0.500000  
4                       0  0.666667  
...                   ...       ...  
11425                   0  0.666667  
11426                  12  0.833333  
11427                   7  0.833333  
11428                   0  0.666667  
11429                   7  0.833333  

[11430 rows x 10 columns]

In [31]:

#entrenamiento (55%) y de prueba (45%)
X_train, X_test, y_train, y_test = train_test_split(df.drop('status_numerico', axis=1), df['status_numerico'], test_size=0.45, random_state=42)

# Divide los datos de entrenamiento en conjuntos de entrenamiento (55%) y de validación (15%)
X_train, X_val, y_train, y_val = train_test_split(X_train, y_train, test_size=0.15/0.55, random_state=42)

# Entrenar el modelo de Random Forest
clf = RandomForestClassifier(random_state=42)
clf.fit(X_train, y_train)

# Realizar predicciones en el conjunto de validación
y_val_pred = clf.predict(X_val)

# Calcular la precisión del modelo en el conjunto de validación
val_accuracy = accuracy_score(y_val, y_val_pred)
print("Accuracy on validation set:", val_accuracy)

# Realizar predicciones en el conjunto de prueba
y_test_pred_random_forrest = clf.predict(X_test)

# Calcular la precisión del modelo en el conjunto de prueba
test_accuracy = accuracy_score(y_test, y_test_pred_random_forrest)
print("Accuracy on test set:", test_accuracy)

y_val_random_forrest = y_val
y_val_pred_random_forrest = y_val_pred

Accuracy on validation set: 0.7895043731778426
Accuracy on test set: 0.7785769828926905


In [32]:
## Gradient boost Machine

In [33]:
from sklearn.model_selection import train_test_split
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.metrics import accuracy_score

# Supongamos que tienes un DataFrame llamado 'df' que contiene tus datos, y una columna 'target' que indica si una URL es phishing o legítima.

# Divide los datos en conjuntos de entrenamiento (55%) y de prueba (45%)
X_train, X_test, y_train, y_test = train_test_split(df.drop('status_numerico', axis=1), df['status_numerico'], test_size=0.45, random_state=42)

# Divide los datos de entrenamiento en conjuntos de entrenamiento (55%) y de validación (15%)
X_train, X_val, y_train, y_val = train_test_split(X_train, y_train, test_size=0.15/0.55, random_state=42)

# Entrenar el modelo de Gradient Boosting Machine
clf = GradientBoostingClassifier(random_state=42)
clf.fit(X_train, y_train)

# Realizar predicciones en el conjunto de validación
y_val_pred = clf.predict(X_val)

# Calcular la precisión del modelo en el conjunto de validación
val_accuracy = accuracy_score(y_val, y_val_pred)
print("Accuracy on validation set:", val_accuracy)

# Realizar predicciones en el conjunto de prueba
y_test_pred_gradient_boost = clf.predict(X_test)

# Calcular la precisión del modelo en el conjunto de prueba
test_accuracy = accuracy_score(y_test, y_test_pred_gradient_boost)
print("Accuracy on test set:", test_accuracy)


Accuracy on validation set: 0.7883381924198251
Accuracy on test set: 0.7828538102643857


In [34]:

# Supongamos que ya tienes los valores verdaderos (y_true) y predichos (y_pred) para tus modelos de Random Forest y Gradient Boosting Machine

# Calcular las métricas para Random Forest
rf_metrics = calculate_metrics(y_val_random_forrest, y_val_pred_random_forrest, 'Random Forest')

# Calcular las métricas para Gradient Boosting Machine
gbm_metrics = calculate_metrics(y_val, y_val_pred, 'Gradient Boosting Machine')

# Concatenar los DataFrames de métricas
metrics_table = pd.concat([rf_metrics, gbm_metrics], ignore_index=True)


In [35]:
metrics_table.head()

Model  True Negative  False Positive  False Negative  \
0              Random Forest            699             147             214   
1  Gradient Boosting Machine            701             145             218   

   True Positive  Precision    Recall   AUC-ROC  
0            655   0.816708  0.753740  0.789991  
1            651   0.817839  0.749137  0.788871

En resumen, tanto el Random Forest como el Gradient Boosting Machine parecen tener un buen rendimiento en la detección de phishing, con altas precisiones, recalls y AUC-ROC.

3. ¿Cuál es el impacto de clasificar un sitio legítimo como phishing?
   Que no le llegue la informacion a una persona que puede ser importante y que ocasione problemas internos y de molestia de los end-users.
4. ¿Cuál es el impacto de clasificar un sitio de phishing como legítimo?
   Que se filtraria informacion de la empresa, hay diferentes niveles pero podria ser desde sus datos personales hasta tarjetas de credito o   informacion sensible.
5. En base a las respuestas anteriores, ¿Qué métrica elegiría para comparar modelos similares
de clasificación de phishing?
    Lastimosamente en este caso ambos me dieron resultados muy similares. Yo buscaria seguir trabajando con ellos hasta encontrar al menos una diferencia significativa para decidirme por uno. Si analizamos sus metricas son muy parecidas.
8. ¿Qué modelo funcionó mejor para la clasificación de phishing? ¿Por qué?
   Ambos funcionaron igual, realmente resaltar en este punto uno sobre el otro seria puro gusto personal ya que sus respuestas son extremadamente similares.
10. Una empresa desea utilizar su mejor modelo, debido a que sus empleados sufren constantes
ataques de phishing mediante e-mail. La empresa estima que, de un total de 50,000 emails, un 15% son phishing. ¿Qué cantidad de alarmas generaría su modelo? ¿Cuántas positivas y cuantas negativas? ¿Funciona el modelo para el BR propuesto? En caso negativo, ¿qué se podría hacer para reducir la cantidad de falsas alarmas?


Utilizando el modelo de Gradient Boosting Machine para determinar cuántas alarmas generaría:

Verdaderos Positivos (TP): 651
Falsos Positivos (FP): 145
Verdaderos Negativos (TN): 701
Falsos Negativos (FN): 218
La cantidad total de alarmas generadas sería la suma de los verdaderos positivos y los falsos positivos:

Total de alarmas generadas = TP + FP = 651 + 145 = 796

De estas alarmas generadas, la cantidad de alarmas positivas (phishing) sería el número de verdaderos positivos:

Cantidad de alarmas positivas = TP = 651

La cantidad de alarmas negativas (no phishing) sería el número de falsos positivos:

Cantidad de alarmas negativas = FP = 145

Ahora, evaluemos si el modelo funciona para el balanceo de precisión y recall (BR) propuesto. Para esto, calcularemos la precisión y el recall del modelo:

Precisión (Precision) = TP / (TP + FP)
Recall = TP / (TP + FN)
Calculamos la precisión:

Precision = 651 / (651 + 145) ≈ 0.818

Calculamos el recall:

Recall = 651 / (651 + 218) ≈ 0.749
